pymagicc/input.py

```diff
@@ -8,6 +8,11 @@
 from six import StringIO
 
 from pymagicc import MAGICC6
+from .definitions import (
+    dattype_regionmode_regions,
+    emissions_units,
+    concentrations_units,
+)
 
```

Comment:
I've basically completely rewritten our `input` module. The idea is that we have a topline object, `MAGICCInput`, which holds our dataframes. It has the metadata attribute as well as a simpler direct access thing than having to always remember how to do pandas indexing. 

Then we have `InputReader`, our superclass for reading files (this could actually be moved into its own file in future). We then subclass `InputReader` to make readers for specific files as necessary e.g. `HIST_CONC_INReader` reads historical concentration files (which are all, in theory, the same format). 

Same idea with writers for `InputWriter` and subclasses. 

The functions `_get_writer` and `_get_reader` choose which class to use to read a file, based off the filename (I think this is the only way to do this, introspecting more heavily is likely a nightmare given internal contradictions in files... happy to change though).

Idea is to extend to reading and writing more file types in future.